packages/contracts/contracts/L1/teleportr/TeleportrDeposit.sol

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.10;
+
+contract TeleportrDeposit {
+    address private owner;
+    uint256 private maxDepositAmount;
+    uint256 private minDepositAmount;
+    uint256 private maxBalance;
+    bool private canReceiveDeposit;
+    uint256 public totalDeposits;
+
+    constructor(
+        uint256 _maxDepositAmount,
+        uint256 _minDepositAmount,
+        uint256 _maxBalance,
+        bool _canReceiveDeposit
+    ) {
+        owner = msg.sender;
+        maxDepositAmount = _maxDepositAmount;
+        minDepositAmount = _minDepositAmount;
+        maxBalance = _maxBalance;
+        canReceiveDeposit = _canReceiveDeposit;
+        totalDeposits = 0;
+        emit OwnerSet(address(0), msg.sender);
+        emit MaxDepositAmountSet(0, _maxDepositAmount);
+        emit MinDepositAmountSet(0, _minDepositAmount);
+        emit MaxBalanceSet(0, _maxBalance);
+        emit CanReceiveDepositSet(_canReceiveDeposit);
+    }
+
+    // Send the contract's balance to the owner
+    function withdrawBalance() public isOwner {
+        uint256 balance = address(this).balance;
+        payable(owner).transfer(balance);
+        emit BalanceWithdrawn(owner, balance);
+    }
+
+    function destroy() public isOwner {
+        emit Destructed(owner, address(this).balance);
+        selfdestruct(payable(owner));
+    }
+
+    // Receive function which reverts if the amount is outside the range
+    // [minDepositAmount, maxDepositAmount], canReceiveDeposit is false, or the
+    // amount would put the contract over its maxBalance.
+    receive()
+        external
+        payable
+        isLowerThanMaxDepositAmount
+        isGreaterThanMinDepositAmount
+        canReceive
+        isLowerThanMaxBalance
+    {
+        uint256 _depositId = totalDeposits + 1;
+        emit EtherReceived(_depositId, msg.sender, msg.value);
+        totalDeposits = _depositId;
+    }
```

Comment:
```suggestion
    {
		totalDeposits += 1;
        emit EtherReceived(totalDeposits, msg.sender, msg.value);
    }
```

Reply:
Sounds good, changed. For my own curiosity, does this approach not use an extra SLOAD?

Reply:
I'd support using foundry if we put this in it's own package (which I think we should do anyways).

Reply:
foundry works in this package with `forge test --hardhat`